File: android/app/src/main/java/com/bluelink/android/domain/PeerIdentityHint.kt
package com.bluelink.android.domain

import java.util.Locale

/** Address evidence selects a candidate for explicit verification; it is never a trust credential. */
object PeerIdentityHint {
    fun bluetooth(address: String?): String? {
        val value = address.orEmpty().trim().replace(":", "").replace("-", "").uppercase(Locale.ROOT)
        return value.takeIf { it.matches(Regex("[0-9A-F]{12}")) && it !in setOf("000000000000", "FFFFFFFFFFFF", "020000000000") }?.let { "bt:$it" }
    }
    fun usbHost(address: String?): String? = address?.takeIf { it.startsWith("usb:BlueLink:") }
        ?.removePrefix("usb:BlueLink:")?.trim()?.takeIf { it.matches(Regex("[0-9a-fA-F]{32}")) }
        ?.let { "usb-host:${it.uppercase(Locale.ROOT)}" }
    fun fromAddress(address: String?): String? = bluetooth(address) ?: usbHost(address)
}

data class IdentityCandidate(val peerId: String, val displayName: String, val publicKey: ByteArray?, val hint: String)

File: windows/BlueLink.App/Security/PeerIdentityHint.cs
using System.Text.RegularExpressions;

namespace BlueLink.Security;

// A hint selects a confirmation candidate. It never authenticates a peer.
public static class PeerIdentityHint
{
    public static string? Bluetooth(string? address)
    {
        var value = (address ?? "").Trim().Replace(":", "").Replace("-", "").ToUpperInvariant();
        return Regex.IsMatch(value, "^[0-9A-F]{12}$") && value is not ("000000000000" or "FFFFFFFFFFFF" or "020000000000")
            ? "bt:" + value : null;
    }
    public static string? UsbSerial(string? serial)
    {
        var value = serial?.Trim();
        return value is { Length: >= 4 and <= 128 } && value.All(c => char.IsAsciiLetterOrDigit(c) || c is '-' or '_' or '.') &&
            value.Any(c => c != '0') ? "usb-serial:" + value.ToUpperInvariant() : null;
    }
}

public sealed record IdentityAssociationHandler(
    Func<string, CancellationToken, Task<IdentityCandidate?>> Find,
    Func<IdentityCandidate, bool> CanAssociate,
    Func<Task> Apply);

public sealed record IdentityCandidate(string PeerId, string DisplayName, byte[]? PublicKey, string Hint);
